### ISSUE TITLE
Preparing Data for Data Nodes

### ISSUE THREAD

**Issue by l45k** [hypha-space/hypha#68]
This issue summarizes the best practices on how to prepare data for Data Nodes. Before diving into the technical details, let's first review the concept of Data Nodes.

In a distributed ML setting with heterogeneous hardware and high latency and low bandwidth connections, we can't follow the beaten paths where each participant receives a similar-sized data partition. Some participants can process more data, due to their computational power, than others. Thus, the Scheduler needs a way to control the amount of data each participant receives and processes.
Additionally, Workers are long-running processes that wait for work to be scheduled and do their computations. Compared to that, a training job is shorter running, and a Job can potentially run on hardware that is not under the control of the person that scheduled the job. With security and efficiency in mind, we need to ensure that training data doesn't contain malicious files and that Workers don't waste resources by downloading the same data.

Therefore, we utilize Data Nodes that host preprocessed and pre-chunked data sets. This gives full control over the data to the Data Scientist and data can be already preprocessed to reduce the computational load on the worker. Additionally, it's the Data Scientists job to ensure that the data and the pre-processing are trustworthy.

To make the process as easy as possible, we decided to use [`safetensors`](https://huggingface.co/docs/safetensors/en/index) as the data format for data sets. It provides libraries for both Python and Rust, and it's the de facto standard for serializing tensor data.

## How to Prepare the Data?

The best way to prepare a data set is the following process:

1. Gain access to the data (downloading it from Hugging Face)
2. Implement a pre-processing pipeline
3. Write chunked `safetensor` files

In Python this can look like the following for preparing the ImageNet dataset from Hugging Face for ResNet training.

```Python
import torch
import safetensors
import os

from datasets import load_dataset

# Load the dataset from Hugging Face. 
dataset = load_dataset("ILSVRC/imagenet-1k", "clean", split="train",token="...", trust_remote_code=True)

data_dir = "datasets/imagenet/train"
# Create a data dir if it doesn't exist
try:
    os.makedirs(data_dir)

counter = 0

# Define a batch function, that preprocesses the data into a batch
def map_to_batch(batch):
    global counter
    # Define an output file name
    filename = f"{data_dir}/train_{counter}.safetensors"
    if not os.path.exists(filename):
        tensors = []
        # Data in ImageNet is stored as images, we need to ensure that they are in RGB format and in the right size
        for i, image in enumerate(batch["image"]):
            # If image isn't RGB, convert it
            if image.mode != "RGB":
                image = image.convert("RGB")
            # Convert the PIL image to a torch tensor
            tensors.append(pil_to_tensor(image.resize((224, 224))))
        safetensors.torch.save_file({"pixel_values": torch.stack(tensors), "labels": torch.tensor(batch["label"])}, filename)
    counter += 1

# Apply the `map_to_batch` function in a mapped way
dataset.map(map_to_batch, batched=True, batch_size=256)
```

Let's break this down. 
We first access the data set hosted on Hugging Face. A token is required since a license agreement must be accepted to use the ImageNet data.

Next, we define a function that batches the whole data set into chunks. We utilize the `map` function on the `dataset` that applies the `map_to_batch` function on a batch of size `256` of the data set. 

The `map_to_batch` function is where the main work happens. The ImageNet files are images that are read in Python with PIL. Not all images are RGB, but we will need them to be RGB; thus, we convert them if needed. Furthermore, the images aren't scaled to the same size. To reduce storage size, we resize the images to 224 by 244 pixels. This is the size we need for ResNet. Finally, we store the images and labels in a `safetensor` file with tensors. One for the images, named `pixel_values` and the other `labels`. The naming here is important, because this is what the model will look for during the training. We will train the model using the `transformers` library. From the docs for the [ResNet model](https://huggingface.co/docs/transformers/model_doc/resnet#transformers.ResNetModel.forward), we know that the inputs to the `forward` function need to be `pixel_values` and `labels`.

The above showcases how we can prepare a dataset for a Data Node. The only thing we now need to do is to copy the data set to the Data Node and add the folder with the training data to its configuration file.
